src/shims/unix/linux/epoll.rs

```diff
@@ -11,24 +13,56 @@ use self::shims::unix::fd::FileDescriptor;
 #[derive(Clone, Debug, Default)]
 struct Epoll {
     /// The file descriptors we are watching, and what we are watching for.
-    file_descriptors: FxHashMap<i32, EpollEvent>,
+    interest_list: BTreeMap<(WeakFileDescriptor, i32), Rc<EpollEvent>>,
+    // ready_list is an Rc because EpollEvents need to hold a reference to update
+    // it.
+    ready_list: Rc<RefCell<BTreeMap<(WeakFileDescriptor, i32), EpollReturn>>>,
+}
+
+/// EpollReturn contains information that will be returned by epoll_wait,
+/// and stored in ready list.
+#[derive(Debug)]
+pub struct EpollReturn {
```

Comment:
This, to me, sounds like a type that should be called `EpollEvent`: it is an event reported by epoll.

The `EpollEvent` type you have maybe should be called `EpollInterest` or so?

Reply:
I have changed the original ``EpollEvent`` to ``EpollInterest``. But since epoll used ``epoll_event`` for both [epoll_ctl](https://man7.org/linux/man-pages/man2/epoll_ctl.2.html) (to register fd) and [epoll_wait](https://man7.org/linux/man-pages/man2/epoll_wait.2.html) (to return notification), would it be a better idea to avoid the ``EpollEvent`` naming and just use ``EpollInterest`` and ``EpollReturn``? Other name for ``EpollReturn`` works too.

Reply:
Why do we have two different types there when they are the same type in epoll?
(There's probably a good reason, I wouldn't be surprised if the C code did some annoying type/field punning here, but I am still curious.)

Maybe `EpollEventInterest` and `EpollEventInstance`? It's quite long, but hopefully also pretty clear.

Reply:
> Why do we have two different types there when they are the same type in epoll?

Because the event field of ``EpollReturn`` and ``EpollInterest`` won't be exactly the same. If we have ``EPOLLIN`` and ``EPOLLOUT`` for ``EpollInterest``, and the file description is only writable, only ``EPOLLOUT`` will be returned. 

Another thing is not all flags will be returned by default, it is mentioned here in [epoll_ctl documentation](https://man7.org/linux/man-pages/man2/epoll_ctl.2.html):
> The events member of the epoll_event structure is a bit mask composed by ORing together zero or more event types, returned by [epoll_wait(2)](https://man7.org/linux/man-pages/man2/epoll_wait.2.html), and input flags, which affect its aren't returned.

So if ``EpollInterest``'s event is (EPOLLIN | EPOLLOUT | EPOLLET), ``EPOLLET`` is classified as ``input flags``, so it would never returned by ``epoll_wait``.